data_pipeline_api/java_api/src/main/java/uk/ramp/access/AccessLoggerImpl.java

```diff
@@ -0,0 +1,66 @@
+package uk.ramp.access;
+
+import java.time.Clock;
+import java.time.Instant;
+import java.util.List;
+import uk.ramp.config.Config;
+import uk.ramp.metadata.MetadataItem;
+
+class AccessLoggerImpl implements AccessLogger {
+  private final List<AccessEntry> accessEntries;
+  private final Clock clock;
+  private final Instant openTimestamp;
+  private final AccessLogWriter writer;
+  private final Config config;
+
+  AccessLoggerImpl(
+      List<AccessEntry> accessEntries,
+      Clock clock,
+      AccessLogWriter writer,
+      Config config,
+      Instant openTimestamp) {
+    this.accessEntries = accessEntries;
+    this.clock = clock;
+    this.openTimestamp = openTimestamp;
+    this.writer = writer;
+    this.config = config;
+  }
+
+  @Override
+  public void logRead(MetadataItem callMetadata, MetadataItem readMetadata) {
+    var accessEntry =
+        ImmutableAccessEntry.builder()
+            .type("read")
+            .timestamp(clock.instant())
+            .callMetadata(callMetadata)
+            .accessMetadata(readMetadata)
+            .build();
+    accessEntries.add(accessEntry);
+  }
+
+  @Override
+  public void logWrite(MetadataItem callMetadata, MetadataItem writeMetadata) {
+    var accessEntry =
+        ImmutableAccessEntry.builder()
+            .type("write")
+            .timestamp(clock.instant())
```

Comment:
How does this timestamp get logged out to access.yaml, as I guess instant is millis since epoch? The upload/download scripts are currently in python, so we need to make sure there's compatibility

Reply:
Good question, and definitely one we should discuss. 

```
timestamp:
  nano: 134173000
  epochSecond: 1594376918
```

This is how it is currently logged out. Happy to change this to a standard format Let me know

Reply:
https://stackoverflow.com/a/45674593

I think if you do this and have it output the "2017-08-14T12:17:47.720Z" format then that's interpretable by the python yaml loader and I assume will also be interpretable by your code without more work than that configuration?

Reply:
That's a nice approach. Will update, thanks.

Reply:
@rtblackwell Made the suggested change and by default it outputs in this format: "2020-07-10T14:08:10.367415Z". Would this be ok, or is it necessary for the seconds/millis be truncated to 3 digits?

Reply:
I think that will work fine
```
import yaml
yaml.safe_load("2020-07-10T14:08:10.367415Z")
datetime.datetime(2020, 7, 10, 14, 8, 10, 367415, tzinfo=datetime.timezone.utc)
``